src/Target.cpp

```diff
@@ -753,6 +774,103 @@ int Target::natural_vector_size(const Halide::Type &t) const {
+            HVX_shared_object, D3D12Compute
+    };
+
+    static const std::vector<Feature> intersectionFeatures = {
+            // These features specify additional instruction sets available on the target architecture.
+            SSE41, AVX, AVX2, FMA, FMA4, F16C, ARMv7s, VSX, AVX512,AVX512_KNL, AVX512_Skylake, AVX512_Cannonlake
+    };
+
+    // I don't know how to merge these features because I don't know what they do. Reviewers, please advise :)
+    static const std::vector<Feature> iDontKnowWhatToDoWithTheseFeaturesPleaseHelp = {
+            CLDoubles, CLHalf, UserContext, Matlab, Profile, CPlusPlusMangling, LargeBuffers, FuzzFloatStores,
+            SoftFloatABI, StrictFloat, LegacyBufferWrappers, CheckUnsafePromises, EmbedBitcode,
+            DisableLLVMLoopVectorize, DisableLLVMLoopUnroll, NoNEON, POWER_ARCH_2_07
+    };
+
+    if (arch != other.arch || bits != other.bits || os != other.os) {
+        user_warning << "runtime targets must agree on platform (arch-bits-os)\n";
+        return false;
+    }
+
+    std::vector<Feature> merged_features{};
+
+    for (const auto &feature : unionFeatures) {
+        if (has_feature(feature) || other.has_feature(feature)) {
+            merged_features.push_back(feature);
+        }
+    }
+
+    for (const auto &feature : intersectionFeatures) {
+        if (has_feature(feature) && other.has_feature(feature)) {
+            merged_features.push_back(feature);
+        }
+    }
+
+    Target output = Target(os, arch, bits, merged_features);
+    if (!fixup_gcd_target(output)) {
+        return false;
+    }
+
+    result = output;
+    return true;
+}
+
+
+bool Target::fixup_gcd_target(Target& target) {
+    // These features do not affect runtime code generation.
+    static const std::vector<Feature> irrelevantFeatures = {
+            JIT, NoAsserts, NoBoundsQuery, NoRuntime, TraceLoads, TraceStores, TraceRealizations
+    };
+
+    // Ensure that irrelevant features are not set
+    for (const auto &feature : irrelevantFeatures) {
+        target.features.reset(feature);
+    }
+
+    // Pick highest CUDA capability. Use fall-through to clear redundant features
+    switch (get_highest_cuda_capability(target)) {
+        case -1:
+            user_warning << "cannot merge cuda-enabled targets without knowing capability levels\n";
+            return false;
+        case 61:
+            target.features.reset(CUDACapability50);
+        case 50:
+            target.features.reset(CUDACapability35);
+        case 35:
+            target.features.reset(CUDACapability32);
+        case 32:
+            target.features.reset(CUDACapability30);
+        default:
+            break;
+    }
+
+    // Check to see that there weren't two or more different HVX flags specified.
+    int number_of_hvx_flags_set = 0;
+    number_of_hvx_flags_set += (int)target.features.test(HVX_64);
+    number_of_hvx_flags_set += (int)target.features.test(HVX_128);
```

Comment:
I believe it should be *exactly* one of {HVX_64, HVX_128}, and at *most* one of the HVX_v* features.

@dsharletg

Reply:
HVX_64 and HVX_128 aren't required, but they should not both be specified.

HVX_v* are treated like CUDA compute capabilities (specifying HVX_v65 also implies HVX_v62).

Reply:
HVX_v* flags are independent of Target::arch. If Target::arch is something other than Hexagon, it still affects the target architecture of offloaded code.

Target::HexagonDMA controls whether or not the Hexagon DMA runtime is included. This is currently only valid if Target::arch is Hexagon.

If none of the HVX_v* flags are set, it defaults to the minimum targeted architecture (v60).

HVX_shared_object looks like it is effectively deprecated, I don't see any uses of it. It used to control whether offloading generated an object or a shared object, but now we always generate a shared object.

FWIW, the motivation for #3673 is to clean up a lot of this (make it easier to deprecate/clean up old deprecated things).

Reply:
Thanks @dsharletg -- I think we can take as a precondition for GCD that the input Targets are valid, though it might be useful to also add a `bool Target::is_valid()` that could be used for sanity checking / assertions. So I'll leave HexagonDMA as a `must_match`.